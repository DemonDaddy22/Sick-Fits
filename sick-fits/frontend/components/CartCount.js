import React from 'react';
import { TransitionGroup, CSSTransition } from 'react-transition-group';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const ANIMATION_DURATION = 500;

const AnimationStyles = styled.span`
    position: relative;
    .count {
        display: block;
        position: relative;
        transition: all ${ANIMATION_DURATION}ms;
        backface-visibility: hidden;
    }
    .count-enter {
        transform: scale(2) rotateX(0.5turn);
    }
    .count-enter-active {
        transform: rotateX(0);
    }
    .count-exit {
        position: absolute;
        top: 0;
        transform: rotateX(0);
    }
    .count-exit-active {
        transform: scale(2) rotateX(0.5turn);
    }
`;

const Dot = styled.div`
    background-color: ${props => props.theme.primary};
    color: ${props => props.theme.white};
    border-radius: 50%;
    padding: 0.5rem;
    line-height: 2rem;
    min-width: 3rem;
    margin-left: 1rem;
    font-weight: 100;
    font-feature-settings: 'tnum';
    font-variant-numeric: tabular-nums;
`;

const CartCount = ({ count }) => <AnimationStyles>
    <TransitionGroup>
        <CSSTransition unmountOnExit className='count' classNames='count' key={count} timeout={{ enter: ANIMATION_DURATION, exit: ANIMATION_DURATION }}>
            <Dot>{count}</Dot>
        </CSSTransition>
    </TransitionGroup>
</AnimationStyles>;

export default CartCount;