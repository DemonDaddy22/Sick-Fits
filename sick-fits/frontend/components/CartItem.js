import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import formatMoney from '../lib/formatMoney';
import RemoveCartItem from './RemoveCartItem';

const CartItemStyles = styled.li`
    padding: 1rem 0;
    border-bottom: 1px solid ${props => props.theme.lightgrey};
    display: grid;
    align-items: center;
    grid-template-columns: auto 1fr auto;

    img {
        margin-right: 1rem;
    }

    h3, p {
        margin: 0;
    }
`;

const CartItem = ({ item }) => <CartItemStyles>
    <img src={item.item.image} alt={item.item.title} width='100' />
    <div className='cart-item-details'>
        <h3>{item.item.title}</h3>
        <p>
            {formatMoney(item.item.price * item.quantity)}
            {' - '}
            <em>
                {item.quantity} &times; {formatMoney(item.item.price)} each
            </em>
        </p>
    </div>
    <RemoveCartItem id={item.id} />
</CartItemStyles>;

CartItem.propTypes = {
    item: PropTypes.object.isRequired
}

export default CartItem;