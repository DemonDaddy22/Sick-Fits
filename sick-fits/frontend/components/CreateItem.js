import React from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Router from 'next/router';
import Form from './styles/Form';
import Error from './ErrorMessage';
import formatMoney from '../lib/formatMoney';

export const CREATE_ITEM_MUTATION = gql`
    mutation CREATE_ITEM_MUTATION(
        $title: String!
        $description: String!
        $price: Int!
        $image: String
        $largeImage: String
    ) {
        createItem(
            data: {
                title: $title
                description: $description
                price: $price
                image: $image
                largeImage: $largeImage
            }
        ) {
            id
        }
    }
`;

export default class CreateItem extends React.Component {

    state = {
        title: '',
        description: '',
        image: '',
        largeImage: '',
        price: 0
    }

    handleChange = e => {
        const { name, type, value } = e.target;
        const val = type === 'number' && val ? parseFloat(value) : value;
        this.setState({ [name]: val });
    }

    handleFileUpload = async e => {
        const files = e.target.files;
        const data = new FormData();
        data.append('file', files[0]);
        data.append('upload_preset', 'sickfits');
        const res = await fetch('https://api.cloudinary.com/v1_1/yelp-camp/image/upload', {
            method: 'POST',
            body: data
        });
        const file = await res.json();
        this.setState({ image: file.secure_url, largeImage: file.eager[0].secure_url });
    }

    handleSubmit = async (e, createItem) => {
        e.preventDefault();
        const res = await createItem();
        Router.push({
            pathname: '/item',
            query: { id: res.data.createItem.id }
        })
    }

    render = () => {
        const { title, description, price, image } = this.state;

        return <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
            {(createItem, { loading, error }) => <Form onSubmit={e => this.handleSubmit(e, createItem)}>
                <Error error={error} />
                <fieldset disabled={loading} aria-busy={loading}>
                    {/* TODO - make input fields bigger and adjust margins */}
                    <label htmlFor='image'>Image</label>
                    <input accept='image/*' type='file' id='image' name='image' placeholder='Upload an Image'
                        onChange={this.handleFileUpload} required />
                    {image && <img width={200} src={image} alt='Upload Preview' />}
                    <label htmlFor='title'>Title</label>
                    <input type='text' id='title' name='title' placeholder='Enter a Title'
                        value={title} onChange={this.handleChange} required />
                    <label htmlFor='price'>Price</label>
                    <input type='number' id='price' name='price' placeholder='Enter a Price'
                        value={price} onChange={this.handleChange} required />
                    <label htmlFor='description'>Description</label>
                    <textarea id='description' name='description' placeholder='Enter a Description'
                        value={description} onChange={this.handleChange} required />
                    <button type='submit'>Submit</button>
                </fieldset>
            </Form>}
        </Mutation>;
    }
}