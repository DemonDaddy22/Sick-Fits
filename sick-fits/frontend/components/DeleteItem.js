import React from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { ALL_ITEMS_QUERY } from '../pages/shop';
export const DELETE_ITEM_MUTATION = gql`
    mutation DELETE_ITEM_MUTATION($id: ID!) {
        deleteItem(where: { id: $id }) {
            id
        }
    }
`;

export default class DeleteItem extends React.Component {

    handleClick = (e, deleteItem) => {
        if (confirm('Are you sure you want to delete this item?')) {
            deleteItem().catch(err => alert(err.message));
        }
    }

    handleUpdate = (cache, payload) => {
        // manually update the cache so that it matches the server
        // read the cache for the items we want
        const data = cache.readQuery({ query: ALL_ITEMS_QUERY });

        // filter the deleted item out of the page
        data.items = data.items.filter(item => item.id !== payload.data.deleteItem.id);

        // put the remaining items back in cache
        cache.writeQuery({ query: ALL_ITEMS_QUERY, data });

    }

    render = () => {

        return <Mutation mutation={DELETE_ITEM_MUTATION} variables={{ id: this.props.id }}
            update={this.handleUpdate}>
            {(deleteItem, { error }) => <button onClick={e => this.handleClick(e, deleteItem)}>
                {this.props.children}
            </button>}
        </Mutation>;
    }
}