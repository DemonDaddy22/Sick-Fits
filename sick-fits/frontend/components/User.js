import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';

export const CURRENT_USER_QUERY = gql`
    query CURRENT_USER_QUERY {
        currentUser {
            id
            email
            name
            permissions
            orders {
                id
            }
            cart {
                id
                quantity
                item {
                    id
                    title
                    description
                    price
                    image
                }
            }
        }
    }
`;

const User = props => <Query query={CURRENT_USER_QUERY} {...props}>
    {payload => props.children(payload)}
</Query>

User.propTypes = {
    children: PropTypes.func.isRequired
}

export default User;