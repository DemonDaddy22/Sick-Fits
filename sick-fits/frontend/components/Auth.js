import { Query } from 'react-apollo';
import styled from 'styled-components';
import { CURRENT_USER_QUERY } from './User';

const TextCenter = styled.p`
    text-align: center;
`;

const Auth = props => <Query query={CURRENT_USER_QUERY}>
    {({ data, loading }) => loading ? <TextCenter>Loading...</TextCenter> :
        !data.currentUser ? <TextCenter>Please sign in to continue</TextCenter> :
            props.children
    }
</Query>;

export default Auth;