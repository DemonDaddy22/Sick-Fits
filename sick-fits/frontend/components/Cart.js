import React from 'react';
import { Query, Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { adopt } from 'react-adopt';
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import CloseButton from './styles/CloseButton';
import SickButton from './styles/SickButton';
import User from './User';
import CartItem from './CartItem';
import getTotalQuantity from '../lib/getTotalQuantity';
import calcTotalPrice from '../lib/calcTotalPrice';
import formatMoney from '../lib/formatMoney';
import TakeMyMoney from './TakeMyMoney';

export const LOCAL_STATE_QUERY = gql`
    query {
        cartOpen @client
    }
`;

export const TOGGLE_CART_MUTATION = gql`
    mutation {
        toggleCart @client
    }
`;

const Composed = adopt({
    user: ({ render }) => <User>{render}</User>,
    toggleCart: ({ render }) => <Mutation mutation={TOGGLE_CART_MUTATION}>{render}</Mutation>,
    localState: ({ render }) => <Query query={LOCAL_STATE_QUERY}>{render}</Query>
});

const Cart = () => <Composed>
    {({ user, toggleCart, localState }) => {
        if (!user.data.currentUser) return null;

        const totalQuantity = getTotalQuantity(user.data.currentUser.cart);

        return <CartStyles open={localState.data.cartOpen}>
            <header>
                <CloseButton onClick={toggleCart} title='Close'>&times;</CloseButton>
                <Supreme>Your Cart</Supreme>
                <p>You have {totalQuantity} item{totalQuantity === 1 ? '' : 's'} in your cart</p>
            </header>
            <ul>
                {user.data.currentUser.cart.map(item => <CartItem key={item.id} item={item}></CartItem>)}
            </ul>
            <footer>
                <p>{formatMoney(calcTotalPrice(user.data.currentUser.cart))}</p>
                <TakeMyMoney>
                    <SickButton>Checkout</SickButton>
                </TakeMyMoney>
            </footer>
        </CartStyles>
    }}
</Composed>;

export default Cart;