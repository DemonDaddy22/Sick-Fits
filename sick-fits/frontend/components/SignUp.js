import React from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Router from 'next/router'
import Form from './styles/Form';
import Error from './ErrorMessage';
import { CURRENT_USER_QUERY } from './User';

export const SIGNUP_MUTATION = gql`
    mutation SIGNUP_MUTATION($name: String!, $email: String!, $password: String!) {
        signup(name: $name, email: $email, password: $password) {
            id
            email
            name
        }
    } 
`;

export default class SignUp extends React.Component {

    state = {
        name: '',
        email: '',
        password: ''
    }

    handleChange = e => this.setState({ [e.target.name]: e.target.value });

    render = () => <Mutation mutation={SIGNUP_MUTATION} variables={this.state} refetchQueries={[{ query: CURRENT_USER_QUERY }]}>
        {(signup, { error, loading }) => <Form method='POST' onSubmit={async e => {
            e.preventDefault();
            // TODO - add catch block to show meaningful error message
            await signup();
            this.setState({ name: '', email: '', password: '' }, () => Router.push({
                pathname: '/'
            }));
        }}>
            <fieldset disabled={loading} aria-busy={loading}>
                <h2>Sign up for an account</h2>
                <Error error={error} />
                <label htmlFor='email'>Email</label>
                <input id='email' type='email' name='email' placeholder='Your email' value={this.state.email} onChange={this.handleChange} />
                <label htmlFor='name'>Name</label>
                <input id='name' type='text' name='name' placeholder='Your name' value={this.state.name} onChange={this.handleChange} />
                <label htmlFor='password'>Password</label>
                <input id='password' type='password' name='password' placeholder='Your password' value={this.state.password} onChange={this.handleChange} />
                <button type='submit'>Sign Up</button>
            </fieldset>
        </Form>}
    </Mutation>;
}