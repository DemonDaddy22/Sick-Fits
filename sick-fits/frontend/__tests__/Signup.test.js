import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { ApolloConsumer } from 'react-apollo';
import { MockedProvider } from 'react-apollo/test-utils';
import SignUp, { SIGNUP_MUTATION } from '../components/SignUp';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser } from '../lib/testUtils';

const currentUser = fakeUser();
const mocks = [{
    request: {
        query: SIGNUP_MUTATION,
        variables: {
            name: currentUser.name,
            email: currentUser.email,
            password: 'foobar'
        }
    },
    result: {
        data: {
            signup: {
                id: 'foo123',
                email: currentUser.email,
                name: currentUser.name,
                __typename: 'User',
            }
        }
    }
},
{
    request: { query: CURRENT_USER_QUERY },
    result: { data: { currentUser } }
}];

const type = (wrapper, name, value) => wrapper.find(`input[name="${name}"]`).simulate('change', { target: { name, value } });

describe('<SignUp />', () => {
    it('renders and matches snapshot', async () => {
        const wrapper = mount(<MockedProvider>
            <SignUp />
        </MockedProvider>);

        expect(toJSON(wrapper.find('form'))).toMatchSnapshot();
    });

    it('calls the mutation', async () => {
        let apolloClient;
        const wrapper = mount(<MockedProvider mocks={mocks}>
            <ApolloConsumer>
                {client => {
                    apolloClient = client;
                    return <SignUp />;
                }}
            </ApolloConsumer>
        </MockedProvider>);

        await wait();
        wrapper.update();

        type(wrapper, 'name', currentUser.name);
        type(wrapper, 'email', currentUser.email);
        type(wrapper, 'password', 'foo123');

        wrapper.update();
        wrapper.find('form').simulate('submit');
        await wait();

        const user = await apolloClient.query({ query: CURRENT_USER_QUERY });
        expect(user.data.currentUser).toMatchObject(currentUser);
    });
});