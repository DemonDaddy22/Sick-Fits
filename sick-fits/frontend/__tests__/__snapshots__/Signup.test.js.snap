// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp /> renders and matches snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 dUkfCz"
  method="POST"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <h2>
      Sign up for an account
    </h2>
    <DisplayError
      error={Object {}}
    />
    <label
      htmlFor="email"
    >
      Email
    </label>
    <input
      id="email"
      name="email"
      onChange={[Function]}
      placeholder="Your email"
      type="email"
      value=""
    />
    <label
      htmlFor="name"
    >
      Name
    </label>
    <input
      id="name"
      name="name"
      onChange={[Function]}
      placeholder="Your name"
      type="text"
      value=""
    />
    <label
      htmlFor="password"
    >
      Password
    </label>
    <input
      id="password"
      name="password"
      onChange={[Function]}
      placeholder="Your password"
      type="password"
      value=""
    />
    <button
      type="submit"
    >
      Sign Up
    </button>
  </fieldset>
</form>
`;
